File: contracts/src/bin/counter.js
// --- Prerequisites ---
// 1. Install the Solana web3 library: `npm install @solana/web3.js`
// 2. You need a funded keypair (the 'payer') on the chosen network (e.g., Devnet).
//    You can airdrop SOL to your keypair address on Devnet/Testnet using the Solana CLI:
//    `solana airdrop 2 <YOUR_WALLET_ADDRESS> --url devnet`
// 3. Run this code using Node.js: `node your_script_name.js`

// Import necessary classes from the Solana web3 library
const {
    Connection,
    Keypair,        // Represents an account keypair (public + private key)
    LAMPORTS_PER_SOL, // Constant for converting SOL to lamports
    PublicKey,      // Represents an account address
    SystemProgram,  // Native program for creating accounts and transferring SOL
    Transaction,    // Represents a Solana transaction
    sendAndConfirmTransaction, // Utility function to send and confirm transactions
} = require("@solana/web3.js");

// --- Configuration ---

// 1. Establish connection to the Solana cluster (Devnet in this case)
// You can replace this with Mainnet-beta, Testnet, or a local validator URL
const connection = new Connection("https://api.devnet.solana.com", "confirmed");
console.log("✅ Connected to Devnet");

// 2. Define the Payer Account
// IMPORTANT: Replace this with your actual secret key array or load it securely.
// NEVER commit secret keys directly into your code in production.
// This keypair needs to have SOL on Devnet to pay for transaction fees and rent.
// Example: Uint8Array.from([ ... your secret key bytes ... ])
// For demonstration, we generate a new one (won't work unless funded externally)
const payer = Keypair.generate(); // In a real scenario, load your funded keypair
console.log(`🔑 Payer Account: ${payer.publicKey.toBase58()}`);
// ** ACTION REQUIRED: You would typically load your keypair here, e.g., from a file **
// const payer = Keypair.fromSecretKey(Uint8Array.from(JSON.parse(require('fs').readFileSync('/path/to/your/keypair.json', 'utf-8'))));

// --- Main Logic ---

const createAccount = async () => {
    try {
        // Request an airdrop to the payer account FOR DEMONSTRATION ONLY
        // In a real app, the payer must already be funded.
        // Airdrops are only available on Devnet/Testnet.
        console.log(`Requesting airdrop for payer ${payer.publicKey.toBase58()}...`);
        const airdropSignature = await connection.requestAirdrop(
            payer.publicKey,
            1 * LAMPORTS_PER_SOL // Request 1 SOL
        );
        await connection.confirmTransaction(airdropSignature, "confirmed");
        console.log("✅ Airdrop successful");

        // 3. Generate a new Keypair for the account we want to create
        const newAccount = Keypair.generate();
        console.log(`🔑 New Account Address: ${newAccount.publicKey.toBase58()}`);

        // 4. Calculate Rent-Exempt Minimum Balance
        // We need to determine the minimum SOL required for the account to be rent-exempt.
        // This depends on the amount of data space allocated (`space`).
        // For this example, we allocate 0 space (just a basic account).
        const space = 0; // Size of data in bytes for the new account
        const rentExemptionMinimum = await connection.getMinimumBalanceForRentExemption(space);
        console.log(`💰 Minimum balance for rent exemption (0 space): ${rentExemptionMinimum} Lamports`);

        // 5. Create the Transaction
        const transaction = new Transaction();

        // 6. Add the "Create Account" instruction
        transaction.add(
            SystemProgram.createAccount({
                fromPubkey: payer.publicKey,          // The account paying for the transaction and rent
                newAccountPubkey: newAccount.publicKey, // The public key of the new account being created
                lamports: rentExemptionMinimum,       // Amount of SOL (in lamports) to transfer to the new account
                space: space,                         // Amount of space (in bytes) to allocate for the account's data
                programId: SystemProgram.programId,   // The owner of the new account (SystemProgram for basic accounts)
            })
        );

        console.log(`⏳ Creating account ${newAccount.publicKey.toBase58()}...`);

        // 7. Sign and Send the Transaction
        // The transaction needs to be signed by the payer and the new account's keypair
        const signature = await sendAndConfirmTransaction(
            connection,
            transaction,
            [payer, newAccount] // Signers: Payer pays, NewAccount proves ownership of the key
        );

        console.log(`✅ Transaction successful! Signature: ${signature}`);
        console.log(`   Account Created: ${newAccount.publicKey.toBase58()}`);

        // 8. Verify Account Creation (Optional)
        const accountInfo = await connection.getAccountInfo(newAccount.publicKey);
        if (accountInfo) {
            console.log("\n🔍 Verification:");
            console.log(`   Owner: ${accountInfo.owner.toBase58()}`); // Should be SystemProgram
            console.log(`   Lamports: ${accountInfo.lamports}`);      // Should match rentExemptionMinimum
            console.log(`   Executable: ${accountInfo.executable}`);  // Should be false
            console.log(`   Data size: ${accountInfo.data.length}`); // Should be 0 (or match 'space')
        } else {
            console.log("❌ Verification failed: Account not found.");
        }

    } catch (error) {
        console.error("❌ Error creating account:", error);
        if (error.logs) {
            console.error("Transaction Logs:", error.logs);
        }
    }
};

// Run the function
createAccount();


// Explanation of the Code:**

// 1.  **Prerequisites & Setup:** Imports the necessary tools from `@solana/web3.js` and sets up a connection to the Solana Devnet. It also defines a `payer` keypair (which needs funding).
// 2.  **Generate New Keypair:** `Keypair.generate()` creates a unique public/private key pair for the account we intend to create on-chain. The public key becomes the account's address.
// 3.  **Calculate Rent:** `connection.getMinimumBalanceForRentExemption(space)` asks the network for the minimum lamports required for an account with the specified `space` (0 bytes in this case) to avoid paying rent.
// 4.  **Create Instruction:** `SystemProgram.createAccount({...})` builds the instruction needed to create the account. It specifies:
//     * `fromPubkey`: Who pays the fees and provides the initial lamports.
//     * `newAccountPubkey`: The address of the account to be created.
//     * `lamports`: The initial balance for the new account (set to the rent-exempt minimum).
//     * `space`: How much data storage to allocate (0 bytes).
//     * `programId`: The **owner** of the new account. Here, `SystemProgram.programId` makes it a standard, basic account owned by the system. To create a data account for *your* program, you would put *your program's ID* here.
// 5.  **Create & Send Transaction:** The instruction is added to a `Transaction`. This transaction is then signed by the `payer` (to authorize fee payment and lamport transfer) and the `newAccount` keypair (to prove control over the new address). `sendAndConfirmTransaction` sends it to the network and waits for confirmation.
// 6.  **Verification:** After confirmation, `connection.getAccountInfo` fetches the details of the newly created account directly from the blockchain, confirming its existence, owner, balance, and other properties.

// This example clearly shows the client-side steps involved in bringing a new account into existence on Solana, highlighting the roles of addresses, rent, ownership (by the System Program in this case), and the transaction proce

File: contracts/src/bin/counter.rs
/*
    The core idea is that everything in Solana is an account. 
    Think of accounts like files in an operating system.


    Data Storage: Accounts are primarily used to store data (state). This could be anything:

A user's wallet balance (SOL tokens).
The code of a deployed program (smart contract).
Data related to an NFT (metadata URI, owner).
State managed by a custom program (like a counter, game state, or DeFi pool information).

Addresses (Public Keys): Every account has a unique address, which is a public key (usually represented as a base58 string).
Ownership: Every account has an owner.

By default, user wallets (accounts holding SOL) are owned by the System Program. 
The System Program is a native Solana program that handles creating new accounts and transferring SOL.
Accounts that store the code of a program are owned by the BPF Loader Program.
Accounts created by your custom program to store its data are typically owned by your program. 
Only the owning program can modify the data within an account it owns 
(except for the System Program which can deduct lamports for rent or transaction fees).

Data: Accounts have a data field, which is just a byte array ([]u8). 
The interpretation of these bytes depends entirely on the program that owns and interacts with the account. 
Programs often use serialization formats like Borsh to structure this data.

Lamports: Accounts hold a balance of lamports (the smallest unit of SOL, 1 SOL=1,000,000,000 lamports). This balance is used for transaction fees and Rent.

Rent: Accounts consume storage space on the Solana cluster. To pay for this, accounts must either:

Pay rent periodically.
Maintain a minimum SOL balance to be considered "rent-exempt". This is the common approach for accounts intended to persist. 
The required balance depends on the size of the data stored in the account.

Executable Flag: Accounts that contain program code have an executable flag set to true. User accounts and data accounts have this set to false.

Stateless Programs: Solana programs themselves are stateless. 
The code stored in a program account doesn't change when it executes. 
All the state it operates on (like a counter's value, user balances in a custom token, etc.) is stored 
in separate data accounts  that are passed into the program during a transaction instruction.

Transactions and Instructions: When you send a transaction, you specify:

Which program you want to execute.
Which accounts will be involved in the instruction (read from or written to).
Any data needed by the instruction itself. 
The Solana runtime uses the list of accounts to potentially schedule transactions in parallel 
if they don't access the same accounts writeably.

Analogy:

Program Account: Like an executable file (.exe or binary) on your computer. It contains instructions but no user data. It's owned by the "operating system" (BPF Loader).
Data Account: Like a data file (.txt, .json, .dat) on your computer. It stores information. It's owned by the program that knows how to read/write it.
User Wallet Account: Like a special system file tracking your core currency, owned by the "operating system" (System Program).
Transaction Instruction: Like running a command: my_program.exe --input user_data.dat --output results.dat. You specify the program and the data files (accounts) it needs to work with.
*/


use solana_program::{
    account_info::{next_account_info, AccountInfo}, // Tools to handle accounts passed into the program
    entrypoint, // Macro to declare the program's entrypoint
    entrypoint::ProgramResult, // Standard result type for programs
    msg, // Macro for logging messages to the chain
    program_error::ProgramError, // Standard error type
    pubkey::Pubkey, // Solana public key type
    program::invoke, // For calling other programs (like System Program) - not used here but common
    system_instruction, // Instructions for the System Program - not used here but common
};
use borsh::{BorshDeserialize, BorshSerialize}; // For serializing/deserializing account data
use std::io::ErrorKind;


// Define the structure of the data we want to store in our data account
#[derive(BorshSerialize, BorshDeserialize, Debug)] // BorshSerialize and BorshDeserialize allow us to easily convert this struct to/from the raw byte array (account_data) stored in the account. why?

pub struct CounterAccount {
    pub counter: u64, // The actual counter value
}

// Define the instructions our program can accept
// In this simple case, we only have one: Increment
// More complex programs would have more variants
// Data for instructions is passed separately from accounts
// (We'll use an empty instruction data for simplicity here)
// enum CounterInstruction {
//     Increment,
//     Decrement, // Example of another instruction
//     Reset { value: u64 } // Example with data
// }


// Program entrypoint function
// Solana runtime calls this function when a transaction targets our program ID
entrypoint!(process_instruction);

// The main logic of our program
pub fn process_instruction(
    program_id: &Pubkey,      // Public key of OUR program account
    accounts: &[AccountInfo], // Array of accounts passed in by the transaction
    _instruction_data: &[u8], // Data passed specific to this instruction (we ignore it here)
) -> ProgramResult { // Must return ProgramResult (Ok or Err)
    msg!("Counter Program Entrypoint");

    // --- 1. Account Validation ---

    // Get the account iterator
    let accounts_iter = &mut accounts.iter();

    // Get the account we expect to store the counter data
    // The client building the transaction must pass this account
    let counter_account = next_account_info(accounts_iter)?;

    // Check 1: Is the counter_account owned by OUR program?
    // This is crucial! Only we should be able to modify the data structure
    // defined by `CounterAccount`.
    if counter_account.owner != program_id {
        msg!("Error: Counter account is not owned by this program");
        return Err(ProgramError::IncorrectProgramId);
    }

    // Check 2: Is the counter_account writable?
    // The transaction must mark this account as writable if we intend to change it.
    if !counter_account.is_writable {
         msg!("Error: Counter account must be writable");
         return Err(ProgramError::InvalidAccountData); // Using this error, adjust as needed
    }

    // --- 2. Instruction Logic ---

    // In a real program, you'd deserialize `_instruction_data` here to figure out
    // *what* action to take (e.g., Increment, Decrement, Reset).
    // For simplicity, we'll assume the only action is Increment.
    // let instruction = CounterInstruction::unpack(_instruction_data)?; // Example

    // --- 3. State Deserialization ---

    // Get the account's data buffer as a slice (mutable borrow because we checked is_writable)
    let mut account_data = counter_account.try_borrow_mut_data()?;

    // Deserialize the byte data into our `CounterAccount` struct
    // Use `try_from_slice` which handles errors gracefully.
    // If the account is new/uninitialized, this might fail.
    let mut counter_state = match CounterAccount::try_from_slice(&account_data) {
         Ok(state) => state,
         Err(e) => {
             // If the error is because the data is empty (uninitialized account),
             // initialize it. Otherwise, propagate the error.
             if e.kind() == ErrorKind::InvalidData || account_data.is_empty() {
                  msg!("Account not initialized. Initializing with counter = 0");
                  CounterAccount { counter: 0 }
             } else {
                 msg!("Error deserializing account data: {:?}", e);
                 return Err(ProgramError::InvalidAccountData);
             }
         }
     };

    // --- 4. Business Logic ---

    // Increment the counter
    counter_state.counter += 1; //where is this counter int defined?
    msg!("Counter incremented. New value: {}", counter_state.counter);

    // --- 5. State Serialization ---

    // Serialize the updated state back into the account's data buffer
     counter_state.serialize(&mut *account_data)?; // The `*` dereferences the mutable borrow RefMut<[u8]>

    msg!("Counter state saved.");
    Ok(()) // Indicate successful execution
}

// Note: This code doesn't handle creating the counter account itself.
// Account creation is usually done by the client (e.g., JavaScript code)
// using the System Program before calling this program's instruction.
// The client would:
// 1. Calculate the required rent-exempt reserve for the size of `CounterAccount`.
// 2. Create a new keypair for the counter account address.
// 3. Send a transaction with `SystemProgram.createAccount` instruction:
//    - Specify the new account's public key.
//    - Allocate space (using `std::mem::size_of::<CounterAccount>()`).
//    - Assign ownership to *this* program's ID (`program_id`).
//    - Transfer enough lamports for rent exemption.
// 4. Then, send a separate transaction calling *this* program's instruction,
//    passing the newly created counter account's public key in the `accounts` array.

